clients/s3/s3.go

```diff
@@ -0,0 +1,235 @@
+/*
+Copyright (c) 2019, VSHN AG, info@vshn.ch
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package s3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	cloudscale "github.com/cloudscale-ch/cloudscale-go-sdk"
+)
+
+// S3EndpointFormat is the endpoint for the S3 API without the region
+const S3EndpointFormat = "https://objects.%s.cloudscale.ch"
+
+// IsErrorNotFound helper function to test for BucketNotFound error
+func IsErrorNotFound(err error) bool {
+	if errResp, ok := err.(*cloudscale.ErrorResponse); ok {
+		return errResp.StatusCode == 404
+	} else if awsErr, ok := err.(awserr.Error); ok {
+		code := awsErr.Code()
+		return code == s3.ErrCodeNoSuchBucket || code == "NotFound"
+	}
+	return false
+}
+
+// Service defines S3 Client operations
+type Service interface {
+	CreateOrUpdateBucket(ctx context.Context, userID, bucketName, region string, cannedACL *string, tags *map[string]string) (*cloudscale.ObjectsUser, error)
+	GetBucketInfo(ctx context.Context, userID, bucketName, region string) (*cloudscale.ObjectsUser, error)
+	DeleteBucket(ctx context.Context, userID, bucketName, region string) error
+}
+
+// Client implements S3 Client
+type Client struct {
+	cloudscaleClient *cloudscale.Client
+}
+
+// NewClient creates a new S3 Client with provided Cloudscale credentials
+func NewClient(ctx context.Context, cloudscaleToken string, httpClient *http.Client) Service {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	c := &Client{
+		cloudscaleClient: cloudscale.NewClient(httpClient),
+	}
+	c.cloudscaleClient.AuthToken = cloudscaleToken
+
+	return c
+}
+
+// CreateOrUpdateBucket creates or updates the supplied S3 bucket with provided
+// specification
+func (c *Client) CreateOrUpdateBucket(ctx context.Context, userID, bucketName, region string, cannedACL *string, tags *map[string]string) (*cloudscale.ObjectsUser, error) {
+	bucketTags := map[string]string{}
+	if tags != nil {
+		bucketTags = *tags
+	}
+	objectUserRequest := &cloudscale.ObjectsUserRequest{
+		DisplayName: bucketName,
+		Tags:        bucketTags,
+	}
+	var objectUser *cloudscale.ObjectsUser
```

Comment:
Our hope is that eventually all Crossplane managed resources will be 1:1 with cloud provider API objects. So in this example, the managed resource would create _only_ a bucket, and not a user. We imagine API objects like this user would be created and managed as a separate managed resource.

That all said, this is totally fine for the moment. If we start modelling (for example) bucket users separately we'll need to arrive on a good pattern for cases like this in which you will almost always want to create a user when you create a bucket. Possibly something like https://github.com/crossplaneio/crossplane/issues/968.